### DISCUSSION TITLE
How to make the same height of matrices?

### DISCUSSION THREAD

**Discussion by sindzicat** [typst/typst#3287] [Q&A] [answered]
Hello!

I need to type the following equation:

![image](https://github.com/typst/typst/assets/33900895/a2d04fcb-fbf2-4a63-acf5-0ddfa18090f2)

I typed:

```typst
$
  mat(
    delim: "[",
    rho_1^t_1, rho_2^t_1, rho_3^t_1, dots.h.c, rho_n^t_1;
    rho_1^t_2, rho_2^t_2, rho_3^t_2, dots.h.c, rho_n^t_2;
    rho_1^t_3, rho_3^t_3, rho_3^t_3, dots.h.c, rho_n^t_3;
    dots.v, dots.v, dots.v, dots.down, dots.v;
    rho_1^t_n, rho_3^t_n, rho_3^t_n, dots.h.c, rho_n^t_n;
  ) dot
  mat(
    delim: "[",
    mu_1; mu_2; mu_3; dots.v; mu_n
  ) =
  mat(
    delim: "[",
    0; 0; 0; dots.v; 0
  )
$
```

I got:

![image](https://github.com/typst/typst/assets/33900895/6ba16e79-48c4-4de2-861b-e34e0c38bc0b)

I tested in [Mathjax](https://www.mathjax.org/#demo):

```
$$
\begin{equation}
    \begin{bmatrix}
        \rho_1^{t_1} & \rho_2^{t_1} & \rho_3^{t_1} & \cdots & \rho_n^{t_1} \\
        \rho_1^{t_2} & \rho_2^{t_2} & \rho_3^{t_2} & \cdots & \rho_n^{t_2} \\
        \rho_1^{t_3} & \rho_3^{t_3} & \rho_3^{t_3} & \cdots & \rho_n^{t_3} \\
        \vdots       & \vdots       & \vdots       & \ddots & \vdots       \\
        \rho_1^{t_n} & \rho_n^{t_n} & \rho_n^{t_n} & \cdots & \rho_n^{t_n} \\
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
        \mu_1 \\ \mu_2 \\ \mu_3 \\ \vdots \\ \mu_n
    \end{bmatrix}
    =
    \begin{bmatrix}
        0 \\ 0 \\ 0 \\ \vdots \\ 0
    \end{bmatrix}
    \label{eq20}
\end{equation}
$$
```
![image](https://github.com/typst/typst/assets/33900895/1233fbbd-87f9-489e-80e0-2352628e7a73)

Now tested on [Katex site](https://katex.org/):

```
$$
  \begin{bmatrix}
      \rho_1^{t_1} & \rho_2^{t_1} & \rho_3^{t_1} & \cdots & \rho_n^{t_1} \\
      \rho_1^{t_2} & \rho_2^{t_2} & \rho_3^{t_2} & \cdots & \rho_n^{t_2} \\
      \rho_1^{t_3} & \rho_3^{t_3} & \rho_3^{t_3} & \cdots & \rho_n^{t_3} \\
      \vdots       & \vdots       & \vdots       & \ddots & \vdots       \\
      \rho_1^{t_n} & \rho_n^{t_n} & \rho_n^{t_n} & \cdots & \rho_n^{t_n} \\
  \end{bmatrix}
  \cdot
  \begin{bmatrix}
      \mu_1 \\ \mu_2 \\ \mu_3 \\ \vdots \\ \mu_n
  \end{bmatrix}
  =
  \begin{bmatrix}
      0 \\ 0 \\ 0 \\ \vdots \\ 0
  \end{bmatrix}
$$
```
![image](https://github.com/typst/typst/assets/33900895/b4b2b503-c5fe-4152-b594-c84e3aaa3448)

Interesting: only on Katex matrices height are the same.

Any way to get the same matrices height in Typst?

**Comment by Julian-Wassmann** [comment, marked as answer]
Hi,

Unfortunately, `Matrix` has no option to do that right now. It can be implemented manually with some effort however.

This function takes in a matrix and a fixed entry height and wraps every entry of that matrix in a block with height set to the given height. The function is built so that it can be used standalone (as a wrapper) and in a matrix show rule. An additional check to avoid infinite recursion is used for the show rule use case.
**Implementation**
```typst
#let fixedHeightMat = (mat, entryHeight) => {
  // return if matrix is empty
  if mat.rows.len() == 0 or mat.rows.at(0).len() == 0 {return mat} 
  let firstEntry = mat.rows.at(0).at(0)
  // return if entries are already wrapped in blocks (using the first entry) to avoid infinite recursion
  if firstEntry.func() == block {return mat} 
  // wrap every entry in block with fixed height
  let entries = mat.rows.map(row => {
    row.map(entry => {
      block(entry, height: entryHeight)
    })
  })
  set align(horizon)
  let delim = if mat.has("delim") {mat.delim} else {"("}
  math.mat(..entries, delim: delim)
}
```

**Show rule use case**
You can apply the show rule multiple times in the document with different heights. But show rules can't be removed right now.
```typst
#show math.mat: mat => fixedHeightMat(mat, 12pt)
```

**Standalone use case**
As an alternative, you can use the function without a show rule.
```typst
$ fixedHeightMat(mat(
  1, 2;
  3, 4;
), #12pt) $
```

Reproduction URL: https://typst.app/project/r1ihpdoUAlpiJw3KHxhDQX

I hope this helps!

**Reply by sindzicat** [reply]
It's possible! O_o It's possible!
Many thanks, @Julian-Wassmann!

**Reply by anmoureaux** [reply]
Hi @Julian-Wassmann, thanks for the solution ! 
However, I noticed that the matrix elements had lost their initial formatting, i.e. they're not in italic anymore (as for standard math expressions). 
The workaround is to replace `block(entry, height: entryHeight)` by `block($entry$, height: entryHeight)` :)
